### ISSUE TITLE
clientEnv is cached for workspace status command

### ISSUE THREAD

**Issue by siddharthab** [bazelbuild/bazel#4222]
Please provide the following information. The more we know about your system and use case, the more easily and likely we can help.

### Description of the problem / feature request / question:
The client environment passed to the BazelWorkspaceStatusModule is cached across runs.
https://github.com/bazelbuild/bazel/blob/d926bc40260549b997a6a5a1e82d9e7999dbb65e/src/main/java/com/google/devtools/build/lib/bazel/BazelWorkspaceStatusModule.java#L106

This means we have to do `bazel clean` whenever any stable value changes as a result of change in environment variables.

### If possible, provide a minimal example to reproduce the problem:
File env_status.sh
```
#!/bin/bash
env | sed -e 's/^/STABLE_/' -e 's/\=/ /'
```

File WORKSPACE is empty.

Demonstration:
```
(bazel build --workspace_status_command=./env_status.sh @bazel_tools//tools/zip:zipper 2>&1 && cat bazel-out/stable-status.txt) | grep STABLE_FOO
# Not found; expected.

# Set a new environment variable and check.
export FOO="Checking environment input."
(bazel build --workspace_status_command=./env_status.sh @bazel_tools//tools/zip:zipper 2>&1 && cat bazel-out/stable-status.txt) | grep STABLE_FOO
# Not found; unexpected.

# Clean cache and try again.
bazel clean
(bazel build --workspace_status_command=./env_status.sh @bazel_tools//tools/zip:zipper 2>&1 && cat bazel-out/stable-status.txt) | grep STABLE_FOO
# Found; expected.

unset FOO
(bazel build --workspace_status_command=./env_status.sh @bazel_tools//tools/zip:zipper 2>&1 && cat bazel-out/stable-status.txt) | grep STABLE_FOO
# Found; unexpected.
```

### Environment info
* Operating System:
macOS 10.13.1

* Bazel version (output of `bazel info release`):
release 0.8.0-homebrew

###  Have you found anything relevant by searching the web?
Nothing related to client environment in relation to stable or volatile status.

**Comment by siddharthab** [comment]
cc @razvanm

**Comment by laszlocsomor** [comment]
@brandjon , not sure if anyone owns this area of Bazel; could you maybe take a look?

**Comment by brandjon** [comment]
I can look in my spare cycles but I won't be able to get to this soon.

**Comment by brandjon** [comment]
Forwarding to dslomov as requested.

**Comment by ulfjack** [comment]
BazelWorkspaceStatusAction.equals is incorrect - it doesn't check for equality of the clientEnv.

**Comment by laszlocsomor** [comment]
@ulfjack : thanks for the analysis!
@dslomov : can you fix it in Q1?

**Comment by justinsb** [comment]
I sent a fix (based on @ulfjack 's suggestion) in #4705
